File: modules/overlay-bubble/src/OverlayBubbleModule.ts
import { requireOptionalNativeModule } from 'expo'
import { Platform } from 'react-native'

// JS surface for the Android chat-head overlay (a floating bubble drawn over other apps).
// The native module exists only on Android; on every other platform `native` is null and
// each call is a safe no-op, so callers don't need their own platform guards. Tapping the
// bubble opens the app at huemobile:// (handled natively via a deep link), so there's no JS
// event to subscribe to — expo-router routes the link to the home screen, which is voice mode.

interface OverlayBubbleNativeModule {
  /** Whether the user has granted the "draw over other apps" (SYSTEM_ALERT_WINDOW) right. */
  hasOverlayPermission: () => boolean
  /** Open the system overlay-permission screen; resolves to the granted state on return. */
  requestOverlayPermission: () => Promise<boolean>
  /** Start the foreground service and show the floating bubble. */
  show: () => Promise<void>
  /** Hide the bubble and stop the foreground service. */
  hide: () => Promise<void>
}

const native = requireOptionalNativeModule<OverlayBubbleNativeModule>('OverlayBubble')

/** True only where the native bubble is available (a release/dev build on Android). */
export const isOverlayBubbleSupported = Platform.OS === 'android' && native != null

export function hasOverlayPermission(): boolean {
  return native?.hasOverlayPermission() ?? false
}

export async function requestOverlayPermission(): Promise<boolean> {
  return (await native?.requestOverlayPermission()) ?? false
}

export async function showBubble(): Promise<void> {
  await native?.show()
}

export async function hideBubble(): Promise<void> {
  await native?.hide()
}

File: modules/overlay-bubble/android/src/main/java/expo/modules/overlaybubble/OverlayBubbleModule.kt
package expo.modules.overlaybubble

import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

// Bridges the floating chat-head overlay to JS. It does three things: report/request the
// "draw over other apps" permission, and start/stop the foreground service that actually
// renders the bubble (see BubbleOverlayService). Tapping the bubble is handled entirely in
// the service via a deep link, so there's no event plumbing here.

class OverlayBubbleModule : Module() {
  // Held while the system overlay-permission screen is open, resolved in OnActivityResult.
  private var pendingPermission: Promise? = null

  override fun definition() = ModuleDefinition {
    Name("OverlayBubble")

    Function("hasOverlayPermission") {
      canDrawOverlays()
    }

    AsyncFunction("requestOverlayPermission") { promise: Promise ->
      if (canDrawOverlays()) {
        promise.resolve(true)
        return@AsyncFunction
      }
      val activity = appContext.activityProvider?.currentActivity
      val context = appContext.reactContext
      if (activity == null || context == null) {
        promise.resolve(false)
        return@AsyncFunction
      }
      pendingPermission = promise
      val intent = Intent(
        Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
        Uri.parse("package:${context.packageName}"),
      )
      activity.startActivityForResult(intent, OVERLAY_PERMISSION_REQUEST)
    }

    OnActivityResult { _, payload ->
      if (payload.requestCode == OVERLAY_PERMISSION_REQUEST) {
        pendingPermission?.resolve(canDrawOverlays())
        pendingPermission = null
      }
    }

    AsyncFunction("show") { promise: Promise ->
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      if (!canDrawOverlays()) {
        promise.reject("E_NO_OVERLAY_PERMISSION", "Overlay permission has not been granted.", null)
        return@AsyncFunction
      }
      val intent = Intent(context, BubbleOverlayService::class.java)
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
        context.startForegroundService(intent)
      } else {
        context.startService(intent)
      }
      promise.resolve(null)
    }

    AsyncFunction("hide") { promise: Promise ->
      val context = appContext.reactContext
      context?.stopService(Intent(context, BubbleOverlayService::class.java))
      promise.resolve(null)
    }
  }

  private fun canDrawOverlays(): Boolean {
    val context = appContext.reactContext ?: return false
    return Settings.canDrawOverlays(context)
  }

  companion object {
    private const val OVERLAY_PERMISSION_REQUEST = 7321
  }
}
